File: leetcode/exercises/3.1.3.medium.koko_eating_bananas_1.py
# -----------------------
# Created     : 28/11/2025
# Last Edited : 29/11/2025 
# Topics      : 
# Big O       :
# Problem Id  : 875. koko eating bananas
# References  : https://www.youtube.com/watch?v=U2SozAs9RzA
# Notes       :
# -----------------------

import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        res = r

        while l <= r:
            k = (r + l) // 2
            hours = 0
            for p in piles:
                hours += math.ceil(p / k)

            if hours <= h:
                res = min(res, k)
                r = k - 1
            else:
                l = k + 1

        return res

File: leetcode/exercises/3.1.3.medium.koko_eating_bananas_1.js
// -----------------------
// Created     : 02/12/2025
// Last Edited : 02/12/2025 
// Topics      : 
// Big O       :
// Problem Id  : 
// References  : https://www.youtube.com/watch?v=U2SozAs9RzA
// Notes       : 
// -----------------------

/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    let l = 1;
    let r = Math.max(...piles);
    let res = 0;
    while (l <= r) {
        k = Math.floor(l + (r - l) / 2)
        let s = 0;
        for (const p of piles) {
            s += Math.ceil(p/k);
        }

        if (s <= h) {
            res = k;
            r = k - 1;
        } else {
            l = k + 1;
        }

    }
    return res;
}

console.log(minEatingSpeed([1, 10, 2, 5, 4, 3], 10));
